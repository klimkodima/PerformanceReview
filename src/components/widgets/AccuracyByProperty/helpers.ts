export const getRangeColor = (score: number): string => {
  switch (true) {
    case 79 < score && score < 101:
      return 'range green';
    case 26 < score && score < 80:
      return 'range yellow';
    default:
      return 'range red';
  }
};
