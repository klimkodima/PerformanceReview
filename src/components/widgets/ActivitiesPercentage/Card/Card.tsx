import { FC, ReactElement } from 'react';
import { useTranslation } from 'react-i18next';

import { CardActivityType } from '../types';

type CardProps = {
  cardData: CardActivityType;
};

const CardView: FC<CardProps> = ({ cardData }): ReactElement => {
  const { t } = useTranslation();
  return (
    <div className='card'>
      <div className='card-row'>
        <label data-testid='totalTimeLabel' className='card-label'>
          {t('activities_percentage.total_time_text')}:
        </label>
        <span data-testid='totalTime' className='card-value'>
          {String(cardData.totalTimeSpend)}h
        </span>
      </div>
      {cardData.labels.map(({ name, value }) => (
        <div key={name} className='card-row'>
          <label className='card-label'>{name}:</label>
          <span className='card-value'>{value}h</span>
        </div>
      ))}
    </div>
  );
};

export default CardView;
