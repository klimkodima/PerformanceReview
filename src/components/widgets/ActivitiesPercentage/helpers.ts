import { getLargestRemainder } from 'src/utils';

import { ActivityLabelType } from 'src/store/ActivitiesPercentage';
import { CardActivityType, LabelDataType } from './types';

export const createPieData = (labels: ActivityLabelType[]): LabelDataType[] => {
  const percentagesList = labels.map(({ percentage }) => percentage);
  // round these percentage values into integers, ensuring that they equal 100% at the end.
  const correctRoundedPercentage = getLargestRemainder(percentagesList, 100);

  return labels.map(({ name }: ActivityLabelType, index) => ({
    name,
    value: correctRoundedPercentage ? correctRoundedPercentage[index] : 0
  }));
};

export const createCardData = (
  labels: ActivityLabelType[],
  totalTimeSpend: number
): CardActivityType => {
  const some = labels.map(({ totalTimeSpend }) => totalTimeSpend / 3600);
  const correctRoundedPercentage = getLargestRemainder(
    some,
    convertSecondsInHours(totalTimeSpend)
  );

  return {
    totalTimeSpend: convertSecondsInHours(totalTimeSpend),
    labels: labels
      .map(({ name }, index) => ({
        name,
        value: correctRoundedPercentage ? correctRoundedPercentage[index] : 0
      }))
      .sort((a: LabelDataType, b: LabelDataType) => b.value - a.value)
  };
};

export const convertSecondsInHours = (secondsTime: number): number =>
  Math.round(secondsTime / 3600);
