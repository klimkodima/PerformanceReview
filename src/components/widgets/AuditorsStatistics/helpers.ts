import { CheckedAuditorsDataType } from 'src/store/Filter/types';
import { AuditorsDataType, AuditorsStatisticsDataType } from './types';

export const createAuditorsStatisticsRows = (
  array: AuditorsStatisticsDataType[],
  checkedAuditors: CheckedAuditorsDataType
): AuditorsDataType[] =>
  array.map(
    ({
      averagePerformance,
      averageValidity,
      hoursCorrespondence,
      supportAndAuditsTime,
      totalPoints,
      auditorId
    }) => ({
      leadName: checkedAuditors[auditorId].teamLeaderName,
      name: checkedAuditors[auditorId].name,
      points: totalPoints,
      performance: averagePerformance,
      validity: averageValidity,
      support: supportAndAuditsTime,
      correspondence: hoursCorrespondence
    })
  );
