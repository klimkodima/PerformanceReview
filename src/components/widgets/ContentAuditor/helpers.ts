import { t } from 'i18next';

import 'src/i18n';

import { inputDataType, outputDataType } from './types';
import { createFormattedLevel } from 'src/utils/commonFunctions';

const getDatePostfix = (date: number, datePeriod: string): string => {
  if (date < 2) {
    return `${date} ${datePeriod}`;
  }

  return `${date} ${datePeriod}s`;
};

const getFullExperiencePeriod = (numberOfMonths: number): string => {
  const countMonths = numberOfMonths % 12;
  const countYears = Math.trunc(numberOfMonths / 12);
  if (countMonths) {
    return `${getDatePostfix(
      countYears,
      t('content_auditor.year')
    )} ${getDatePostfix(countMonths, t('content_auditor.month'))}`;
  } else {
    return `${getDatePostfix(countYears, t('content_auditor.year'))}`;
  }
};

export const getExperience = (numberOfMonths: number): string => {
  if (numberOfMonths < 2) {
    return `${numberOfMonths} ${t('content_auditor.month')}`;
  } else if (numberOfMonths > 12) {
    return getFullExperiencePeriod(numberOfMonths);
  } else {
    return `${numberOfMonths} ${t('content_auditor.month')}s`;
  }
};

export const getOutputData = ({
  key,
  value
}: inputDataType): outputDataType => {
  switch (key) {
    case t('content_auditor.full_name'):
      return { label: t('content_auditor.name'), text: `${value}`, order: 0 };
    case t('content_auditor.level'):
      return {
        label: t('content_auditor.current_level'),
        text: `${createFormattedLevel(value as string)}`,
        order: 2
      };
    case t('content_auditor.team_lead_name'):
      return {
        label: t('content_auditor.team_lead'),
        text: `${value}`,
        order: 3
      };
    case t('content_auditor.experience_in_month'):
      return {
        label: t('content_auditor.experience'),
        text: `${getExperience(value as number)}`,
        order: 4
      };
    case t('content_auditor.team_name'):
      return {
        label: t('content_auditor.team'),
        text: `${value}`,
        order: 1
      };
    default:
      return { label: '', text: '', order: 0 };
  }
};
