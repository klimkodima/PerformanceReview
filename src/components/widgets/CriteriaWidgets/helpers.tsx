import { t } from 'i18next';
import 'src/i18n';

export const getValueType = (titleName: string, value: number): string => {
  if (titleName === t('criteria_widgets.averagePerformance')) {
    return value === 1
      ? t('criteria_widgets.task_h')
      : t('criteria_widgets.tasks_h');
  } else if (titleName === t('criteria_widgets.supportAndAuditsTime')) {
    return 'h';
  } else return '%';
};
