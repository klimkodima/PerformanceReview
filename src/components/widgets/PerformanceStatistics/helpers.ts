import { t } from 'i18next';
import { isEmpty } from 'lodash';

import { CheckedAuditorsDataType } from 'src/store/Filter/types';
import {
  PerfStatOmitType,
  PerfStatType
} from 'src/store/PerformanceStatistic/types';
import { UserPointsType } from './types';

export const getAllDates = (
  performanceStatisticData: PerfStatType[]
): string[] =>
  performanceStatisticData.map(({ from }) => from.slice(5).replace(/-/g, '.'));

export const getAllPoints = (
  performanceStatisticData: PerfStatType[],
  checkedAuditors: CheckedAuditorsDataType
): UserPointsType[] => {
  const dataAPI: UserPointsType[] = [
    {
      name: t('performance_statistics.average_senior_auditor'),
      data: []
    },
    {
      name: t('performance_statistics.average_middle_auditor'),
      data: []
    },
    {
      name: t('performance_statistics.average_junior_auditor'),
      data: []
    }
  ];

  if (!isEmpty(checkedAuditors)) {
    for (const key in checkedAuditors) {
      dataAPI.push({ name: checkedAuditors[key].name, data: [] });
    }
  }

  if (
    isEmpty(checkedAuditors) &&
    performanceStatisticData[0]?.usersPointsDots
  ) {
    dataAPI.push({
      name: performanceStatisticData[0].usersPointsDots[0].userName,
      data: []
    });
  }

  performanceStatisticData.forEach(
    ({
      juniorsPoints,
      middlesPoints,
      seniorsPoints,
      usersPointsDots
    }: PerfStatOmitType) => {
      for (const key of dataAPI) {
        if (
          key.name.includes(t('performance_statistics.average_junior_auditor'))
        ) {
          key.data.push(juniorsPoints);
        } else if (
          key.name.includes(t('performance_statistics.average_middle_auditor'))
        ) {
          key.data.push(middlesPoints);
        } else if (
          key.name.includes(t('performance_statistics.average_senior_auditor'))
        ) {
          key.data.push(seniorsPoints);
        }
        if (usersPointsDots) {
          for (const userInfo of usersPointsDots) {
            if (key.name === userInfo.userName) {
              key.data.push(userInfo.userPoints);
            }
          }
        }
      }
    }
  );

  return dataAPI;
};

export const getSubtitle = (from: string, to: string): string => {
  const date1 = new Date(from);
  const date2 = new Date(to);
  const oneDay = 1000 * 60 * 60 * 24;
  const diffInTime = date2.getTime() - date1.getTime();
  const diffInDays = Math.round(diffInTime / oneDay);

  if (diffInDays === 0) return t('performance_statistics.day');
  if (diffInDays === 6) return t('performance_statistics.week');
  if (diffInDays === 29) return t('performance_statistics.month');
  return t('performance_statistics.year');
};
