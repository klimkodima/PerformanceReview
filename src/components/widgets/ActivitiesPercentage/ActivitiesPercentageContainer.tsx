import { ReactElement, useEffect } from 'react';

import { useDispatch, useSelector } from 'react-redux';

import { useTranslation } from 'react-i18next';

import ActivitiesPercentage from './ActivitiesPercentage';
import {
  fetchAuditActivities,
  selectAuditActivities
} from 'src/store/ActivitiesPercentage';
import { selectPickedDate } from 'src/store/DateFilter';
import { createCardData, createPieData } from './helpers';

import { PieOptionType } from './types';

const ActivitiesPercentageContainer = (): ReactElement => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const pickedDate = useSelector(selectPickedDate);
  useEffect(() => {
    dispatch(fetchAuditActivities());
  }, [pickedDate]);

  const { totalTimeSpend, labels } = useSelector(selectAuditActivities);

  const pieData = createPieData(labels);
  const cardData = createCardData(labels, totalTimeSpend);

  const option: PieOptionType = {
    animation: false,
    labelLine: {
      color: 'black'
    },
    label: {
      show: true,
      color: 'black',
      fontWeight: 500,
      fontSize: '1.3rem',
      formatter: '{b}: {d}%'
    },
    series: [
      {
        name: t('activities_percentage.pie_title_text'),
        type: 'pie',
        radius: '57%',
        data: pieData,
        labelLine: {
          lineStyle: {
            color: 'black'
          }
        },
        emphasis: {
          disabled: true
        },
        top: '0'
      }
    ]
  };

  return <ActivitiesPercentage pieData={option} cardData={cardData} />;
};

export default ActivitiesPercentageContainer;
