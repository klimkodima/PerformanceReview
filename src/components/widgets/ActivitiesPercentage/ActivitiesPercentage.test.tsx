import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import '@testing-library/jest-dom/extend-expect';

import i18n from 'src/i18n';
import store from 'src/store';
import {
  AuditActivitiesType,
  setAuditActivities
} from 'src/store/ActivitiesPercentage';
import ActivitiesPercentageContainer from './ActivitiesPercentageContainer';
import { Card } from './Card';
import { createCardData } from './helpers';

const mockData: AuditActivitiesType = {
  totalTimeSpend: 225600,
  labels: [
    {
      name: 'Audits',
      totalTimeSpend: 90000,
      percentage: 39.8936170212766
    },
    {
      name: 'Meetings',
      totalTimeSpend: 55800,
      percentage: 24.73404255319149
    },
    {
      name: 'Others',
      totalTimeSpend: 37800,
      percentage: 16.75531914893617
    },
    {
      name: 'Support',
      totalTimeSpend: 42000,
      percentage: 18.617021276595743
    }
  ]
};

const cardData = createCardData(mockData.labels, mockData.totalTimeSpend);

describe('Card  Component', () => {
  describe('Matches snapshots', () => {
    it('Matches not zero snapshot', () => {
      const card = renderer
        .create(
          <I18nextProvider i18n={i18n}>
            <Card cardData={cardData} />
          </I18nextProvider>
        )
        .toJSON();
      expect(card).toMatchSnapshot();
    });
  });
  describe('Matches total label and value', () => {
    it('Matches total value', () => {
      render(
        <I18nextProvider i18n={i18n}>
          <Card cardData={cardData} />
        </I18nextProvider>
      );
      expect(screen.getByTestId('totalTime')).toHaveTextContent(
        String(cardData.totalTimeSpend)
      );
    });
    it('Matches total label', () => {
      render(
        <I18nextProvider i18n={i18n}>
          <Card cardData={cardData} />
        </I18nextProvider>
      );
      expect(screen.getByTestId('totalTimeLabel')).toHaveTextContent(
        'Total time:'
      );
    });
  });
  describe('Matches activity labels and values', () => {
    it('Matches activity labels', () => {
      const { container } = render(
        <I18nextProvider i18n={i18n}>
          <Card cardData={cardData} />
        </I18nextProvider>
      );
      expect(container.getElementsByClassName('card-label')).toHaveLength(
        cardData.labels.length + 1
      );
    });
    it('Matches activity values', () => {
      const { container } = render(
        <I18nextProvider i18n={i18n}>
          <Card cardData={cardData} />
        </I18nextProvider>
      );
      expect(container.getElementsByClassName('card-value')).toHaveLength(
        cardData.labels.length + 1
      );
    });
  });
  describe('Matches sorted activities', () => {
    it('Matches max activity', () => {
      const { container } = render(<Card cardData={cardData} />);
      expect(container.getElementsByClassName('card-value')).toHaveLength(5);
      expect(
        container.getElementsByClassName('card-value')[1]
      ).toHaveTextContent(String(cardData.labels[0].value) + 'h');
    });
    it('Matches min actievity', () => {
      const { container } = render(<Card cardData={cardData} />);
      expect(container.getElementsByClassName('card-value')).toHaveLength(5);
      expect(
        container.getElementsByClassName('card-value')[2]
      ).toHaveTextContent(String(cardData.labels[1].value) + 'h');
    });
  });
});

describe('ActivitiesPercentageContainer', () => {
  test('Matches snapshot', () => {
    store.dispatch(setAuditActivities(mockData));

    const { container, getByTestId } = render(
      <Provider store={store}>
        <ActivitiesPercentageContainer />
      </Provider>
    );
    expect(getByTestId('totalTime')).toHaveTextContent('63h');
    expect(container.getElementsByClassName('card-value')[1]).toHaveTextContent(
      String(cardData.labels[0].value) + 'h'
    );
    expect(container.getElementsByClassName('card-value')[2]).toHaveTextContent(
      String(cardData.labels[1].value) + 'h'
    );
    expect(container.getElementsByClassName('card-value')[3]).toHaveTextContent(
      String(cardData.labels[2].value) + 'h'
    );
  });

  test('Matches snapshot with 33 percentages', () => {
    const mock33PercentageData: AuditActivitiesType = {
      totalTimeSpend: 32400,
      labels: [
        {
          name: 'Audits',
          totalTimeSpend: 10800,
          percentage: 33.333333
        },
        {
          name: 'Meetings',
          totalTimeSpend: 10800,
          percentage: 33.333333
        },
        {
          name: 'Support',
          totalTimeSpend: 10800,
          percentage: 33.333333
        }
      ]
    };

    const card33PercentageData = createCardData(
      mock33PercentageData.labels,
      mock33PercentageData.totalTimeSpend
    );

    store.dispatch(setAuditActivities(mock33PercentageData));

    const { container, getByTestId } = render(
      <Provider store={store}>
        <ActivitiesPercentageContainer />
      </Provider>
    );
    expect(getByTestId('totalTime')).toHaveTextContent('9h');
    expect(container.getElementsByClassName('card-value')[1]).toHaveTextContent(
      String(card33PercentageData.labels[0].value) + 'h'
    );
    expect(container.getElementsByClassName('card-value')[2]).toHaveTextContent(
      String(card33PercentageData.labels[1].value) + 'h'
    );
    expect(container.getElementsByClassName('card-value')[3]).toHaveTextContent(
      String(card33PercentageData.labels[2].value) + 'h'
    );
  });
});
