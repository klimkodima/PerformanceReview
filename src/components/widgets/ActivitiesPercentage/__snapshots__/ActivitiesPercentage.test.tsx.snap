// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card  Component Matches snapshots Matches not zero snapshot 1`] = `
<div
  className="card"
>
  <div
    className="card-row"
  >
    <label
      className="card-label"
      data-testid="totalTimeLabel"
    >
      Total time
      :
    </label>
    <span
      className="card-value"
      data-testid="totalTime"
    >
      63
      h
    </span>
  </div>
  <div
    className="card-row"
  >
    <label
      className="card-label"
    >
      Audits
      :
    </label>
    <span
      className="card-value"
    >
      26
      h
    </span>
  </div>
  <div
    className="card-row"
  >
    <label
      className="card-label"
    >
      Meetings
      :
    </label>
    <span
      className="card-value"
    >
      16
      h
    </span>
  </div>
  <div
    className="card-row"
  >
    <label
      className="card-label"
    >
      Support
      :
    </label>
    <span
      className="card-value"
    >
      11
      h
    </span>
  </div>
  <div
    className="card-row"
  >
    <label
      className="card-label"
    >
      Others
      :
    </label>
    <span
      className="card-value"
    >
      10
      h
    </span>
  </div>
</div>
`;
